refactor(common): change encoding function in httpparams

edodusi opened the pull request:
HttpParams uses custom encoding function "standardEncoding" to encode query string preserving specific charachters. This refactoring aims to improve performances and code quality of that function by using a RegExp and a Map object instead of a chain of replace functions.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Parameters string passes through `encodeURIComponent` and then into a chain of `replace` that un-encodes some characters.
Although it has been pointed out that this behaviour is non-standard and hard to understand for some developers (cfr. #19710) this commit doesn't change the function's output but only improves its internals.

## What is the new behavior?

-

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

edodusi committed:
refactor(common): change encoding function in httpparams

HttpParams uses custom encoding function "standardEncoding" to encode query string preserving specific charachters. This refactoring aims to improve performances and code quality of that function by using a RegExp and a Map object instead of a chain of replace functions.

edodusi committed:
Merge branch 'master' into better-httpparams-encoding

petebacondarwin commented on packages/common/http/src/params.ts:
```diff
@@ -78,17 +78,22 @@ function paramParser(rawParams: string, codec: HttpParameterCodec): Map<string,
-      .replace(/%3D/gi, '=')
-      .replace(/%3F/gi, '?')
-      .replace(/%2F/gi, '/');
+  const tmap = new Map<string, string>([
+    ['40', '@'],
+    ['3A', ':'],
+    ['24', '$'],
+    ['2C', ','],
+    ['3B', ';'],
+    ['2B', '+'],
+    ['3D', '='],
+    ['3F', '?'],
+    ['2F', '/'],
+  ]);
```
There is no need to use `Map` here you can just use a normal JS object:

```ts
const map = {
  '40': '@',
  '3A': ':',
  '24': '$',
  '2C': ',',
  '3B': ';',
  '2B': '+',
  '3D': '=',
  '3F': '?',
  '2F': '/',
};
```

Also this could be moved to the top level of the file to avoid initializing it on every call to `standardEncoding()`.

petebacondarwin commented on packages/common/http/src/params.ts:
```diff
@@ -78,17 +78,22 @@ function paramParser(rawParams: string, codec: HttpParameterCodec): Map<string,
+    ['3F', '?'],
+    ['2F', '/'],
+  ]);
+  return encodeURIComponent(v).replace(/%(\d[a-z0-9])/gi, (_, t) => tmap.get(t) || `%${t}`);
```
No need to recreate the string since it is passed as the first parameter.

```suggestion
  return encodeURIComponent(v).replace(/%(\d[a-z0-9])/gi, (s, t) => tmap.get(t) ?? s);
```

petebacondarwin reviewed (changes requested):
Thanks for this PR - I like the idea to avoid searching the whole string for each of the possible replacements. I think we can tighten it up even more.

edodusi commented:
Thanks @petebacondarwin, tried yout suggested refactoring and it's fine (also added a bit of typings).
Unfortunately I deleted the fork repo from which I opened this PR (because 2 years 😅), so I don't think I can add any more commits to this PR now. What should I do?

petebacondarwin commented:
The simplest solution will be to reproduce the PR on a new branch and close this PR.

edodusi commented:
Yep. Ok, closing this one, will open a new PR and tag you in a comment, since I don't think I can assign or add a reviewer.
Thanks again!
